README.md
# FuncProgHW1

Functional Programming, Assignment 1.

Implements *Zeller's congruence* to solve the *Counting Sundays* problem.

#### Equivalent ES7 implementation

```javascript
function dayOfWeek(day, month, year) {
  const yearOfCentury = year % 100;
  const zeroBasedCentury = Math.floor(year / 100);
  const mappedMonth = [13, 14, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12][month];

  return (
    day + 
    Math.floor((13 * (mappedMonth + 1)) / 5) + 
    yearOfCentury + 
    Math.floor(yearOfCentury / 4) + 
    Math.floor(zeroBasedCentury / 4) + 
    (5 * zeroBasedCentury)
  ) % 7;
}
```

## Installation and usage

1. Run *cabal* build engine: `$ cabal install -j`.
2. Run `$ .cabal-sandbox/bin/ZellersCongruence` to start the application.

## Running HLint

1. Run `$ cabal install hlint`.
2. Run `$ .cabal-sandbox/bin/hlint .` to execute the *HLint*.

## Answer for the math question

Quoting from [Wikipedia.org](https://en.wikipedia.org/wiki/Leap_year):

> ... in the Gregorian calendar, each leap year has 366 days instead of the usual 365, by extending February to 29 days rather than the common 28. These extra days occur in years which are multiples of four (with the exception of years divisible by 100 but not by 400).

Each week contains 7 days. Without leap days, 400 years would contain (365 * 400) days, which would
be equal to **146.000** days.

There should be 24 leap days in a century, if the destination year is not divisible by 400, which adds (24 * 3) days, which would be equal to **72**.

One of the centuries found in 400 years interval has **25** leap days since it is divisible by 400.

That makes **146.097** days, number of days in 400 years is a multiple of 7.

Picking one day is bounded to the same probability, all days are equally possible since this is a uniform distribution.

## License

MIT with additional statements. See license file for more information.


ChangeLog.md
# Revision history for zellerscongruence

## 0.1.0

* Release without unit tests, completely experimental.


main.hs
{- |
Module      :  ZellersCongruence
Description :  Application start point.
Copyright   :  (c) 2018 Buğra Ekuklu (chatatata)
License     :  See LICENSE.md

Maintainer  :  ekuklu[at]icloud.com
Stability   :  experimental
Portability :  portable

Implements *Zeller's congruence* in order to calculated the day of the week on given day, month
and year in Gregorian calendar.
-}
module ZellersCongruence where
  import Prelude

  -- | Size of a century.
  centurySize :: Integer
  centurySize = 100

  -- | Data type representing day of a week.
  data DayOfWeek = 
    Saturday | Sunday | Monday | Tuesday | Wednesday | Thursday | Friday  deriving (Enum)

  -- | Calculates day of a week on a Gregorian calendar by using *Zeller's congruence* method.
  dayOfWeek :: Integer -- ^ Day of the month, e.g. 21.
            -> Integer -- ^ Month of the year.
            -> Integer -- ^ Year.
            -> Integer -- ^ Calculated day of the week.
  dayOfWeek day month year = 
      calculatedDay `mod` 7 where
    yearOfCentury :: Integer
    yearOfCentury = year `mod` centurySize

    zeroBasedCentury :: Integer
    zeroBasedCentury = floor(fromInteger(year) / fromInteger(centurySize))

    calculatedDay :: Integer
    calculatedDay = month + yearOfCentury + (realToFrac(zeroBasedCentury) / realToFrac(4.00))


src/Main.hs
{- |
Module      :  Main
Description :  Application start point.
Copyright   :  (c) 2018 Buğra Ekuklu (chatatata)
License     :  See LICENSE.md

Maintainer  :  ekuklu[at]icloud.com
Stability   :  experimental
Portability :  portable

Implements *Zeller's congruence* in order to calculated the day of the week on given day, month
and year in Gregorian calendar.
-}
module Main where

import Prelude

-- | Application start point.
main :: IO ()
main = do print (sundays1 1901 2000)
          print (sundays2 1901 2000)
          print (sundays1tr 1901 2000)

-- | Size of a century.
centurySize :: Integer
centurySize = 100

-- | Width of a leap.
leapWidth :: Integer
leapWidth = 4

-- | Size of a week in calendar.
weekSize :: Integer
weekSize = 7

-- | Calculates day of a week on a Gregorian calendar by using *Zeller's congruence* method.
dayOfWeek :: Integer -- ^ Year.
          -> Integer -- ^ Month of the year.
          -> Integer -- ^ Day of the month, e.g. 21.
          -> Integer -- ^ Calculated day of the week.
dayOfWeek year realMonth day = calculatedDay `mod` weekSize
  where
    month :: Integer
    month = realMonth - 1
    yearOfCentury :: Integer
    yearOfCentury = year `mod` centurySize
    zeroBasedCentury :: Integer
    zeroBasedCentury = floor (fromIntegral year / fromIntegral centurySize)
    mappedMonth :: Integer
    mappedMonth = [13, 14, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]!!fromIntegral month
    calculatedDay :: Integer
    calculatedDay =
      day +
      floor (fromIntegral (13 * (mappedMonth + 1)) / 5.00) +
      yearOfCentury +
      floor (fromIntegral yearOfCentury / 4.00) +
      floor (fromIntegral zeroBasedCentury / 4.00) +
      (5 * zeroBasedCentury)

-- | Calculates the number of months starting with sundays in given time interval of years.
sundays1 :: Integer -- ^ Starting year of the interval.
         -> Integer -- ^ Ending year of the interval.
         -> Integer -- ^ Number of sundays between given years.
sundays1 start end = sundays' start 1
  where
    sundays' :: Integer -> Integer -> Integer
    sundays' y m
      | y > end   = 0
      | otherwise = if dayOfWeek y m 1 == 1 then rest + 1 else rest
      where
        nextY = sundays' (y + 1) 1
        nextM = sundays' y (m + 1)
        rest  = if m < 12 then nextM else nextY

-- | Tail recursive variant of `sundays1 (Integer, Integer) -> Integer`.
sundays1tr :: Integer -- ^ Starting year of the interval.
           -> Integer -- ^ Ending year of the interval.
           -> Integer -- ^ Number of sundays between given years.
sundays1tr start end = sundays' start 1 0
  where
    sundays' :: Integer -> Integer -> Integer -> Integer
    sundays' y m acc
      | y > end   = acc
      | otherwise = if dayOfWeek y m 1 == 1 then inc else pass
      where
      nextY = sundays' (y + 1) 1
      nextM = sundays' y (m + 1)
      inc = if m < 12 then nextM (acc + 1) else nextY (acc + 1)
      pass = if m < 12 then nextM acc else nextY acc

-- | Finds out if the given year has a leap day.
leap :: Integer -- ^ Year.
     -> Bool -- ^ Truthy value if year has a leap day, otherwise false.
leap year = isForth && not isCenturyStart || isForthCenturyStart
  where
    isForth :: Bool
    isForth = (year `mod` leapWidth) == 0
    isCenturyStart :: Bool
    isCenturyStart = (year `mod` centurySize) == 0
    isForthCenturyStart :: Bool
    isForthCenturyStart = (year `mod` (centurySize * leapWidth)) == 0

-- | Calculates days in given month-year pair.
daysInMonth :: Integer -- ^ Month of the year.
            -> Integer -- ^ Year.
            -> Integer -- ^ Number of days.
daysInMonth month year
  | month == 2  = if leap year then 29 else 28
  | month == 4 || 
    month == 6 || 
    month == 9 || 
    month == 11 = 30
  | otherwise   = 31

-- | Second variant of function `sundays1 (Integer, Integer) -> Integer`, using week days in order
-- calculate number of sundays between given years.
sundays2 :: Integer -- ^ Starting year of the interval.
         -> Integer -- ^ Ending year of the interval.
         -> Integer -- ^ Number of sundays between given years.
sundays2 start end = sundays' start 1 2
  where
    sundays' :: Integer -> Integer -> Integer -> Integer
    sundays' year month weekDay
      | year > end   = 0
      | otherwise = if nextWeekday `mod` 7 == 0 then rest + 1 else rest
      where
        nextY :: Integer
        nextY = sundays' (year + 1) 1 nextWeekday
        nextM :: Integer
        nextM = sundays' year (month + 1) nextWeekday
        nextWeekday :: Integer
        nextWeekday = weekDay + (daysInMonth month year `mod` weekSize)
        rest :: Integer
        rest  = if month < 12 then nextM else nextY
